### PR TITLE
[CASSANDRA-16378] Expose all client options via system_views.clients and nodetool clientstats

### PR DESCRIPTION
**Pull request by rtib**

### CODE REVIEW

File: src/java/org/apache/cassandra/transport/ConnectedClient.java [hunk 1]
```diff
@@ -128,10 +137,20 @@ private SslHandler sslHandler()
...
     {
+        String clientOptionsJSON;
+        ObjectMapper objMapper = new ObjectMapper();
+        try
+        {
+            clientOptionsJSON = objMapper.writeValueAsString(clientOptions().get());
+        } catch(JsonProcessingException e)
+        {
+            clientOptionsJSON = UNDEFINED;
+        }
```
Comment (blerer) [inline src/java/org/apache/cassandra/transport/ConnectedClient.java]: We do not normally use json libraries for that type of things. It is also more readable to us Guava Joiner:
` Joiner.on(',').withKeyValueSeparator("=").join(clientOptions().get())`

**Review by blerer — COMMENTED** [review]
We do not have the mean to really test the new options but we should be able to partially test the output of the `ClientsTable` queries by verifying only the part that we can predict. Could you create a unit test similar to `SystemPropertiesTableTest`? The call to `executeNet` will open one connection to the server.

**Comment by rtib** [comment]
I've added a ClientsTableTest which is querying the virtual table and asserting some values.

First I've added assertion for driver_name and driver_version columns, which values should be contained in the client_options, however, this depends on the behaviour of StartupMessage rather than Clients itself, thus testing this is violating the SUT boundary. I don't know if it is a good idea to keep those ones.

File: test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        for (Row r : result)
+        {
+            Assert.assertEquals(r.getInet("address"), InetAddress.getLoopbackAddress());
+            Assert.assertNotNull(r.getInt("port"));
+            Assert.assertTrue(r.getInt("port") > 0);
+            Assert.assertNotNull(r.getMap("client_options", String.class, String.class));
+            // the following are questionable if they belong here
+            Assert.assertEquals(r.getString("hostname"), "localhost");
+            Assert.assertEquals(r.getString("driver_name"), r.getMap("client_options", String.class, String.class).get("DRIVER_NAME"));
+            Assert.assertEquals(r.getString("driver_version"), r.getMap("client_options", String.class, String.class).get("DRIVER_VERSION"));
```
Comment (blerer) [inline test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java]: The assertions surrounding the map can be be simplified by using AspectJ assertions (`org.assertj.core.api.Assertions`). Like: `assertThat(map).hasEntrySatisfying(...)`
I think that we can also test `request_count`.

**Review by blerer — COMMENTED** [review]
Nice work :-) I just have some minor nits.

**Comment by rtib** [comment]
Asserting request_count to some positive, as it has different values in the two rows. Changed map assertion as suggested.

**Comment by rtib** [comment]
Squashed and rebased to current trunk.

File: test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            Assert.assertTrue(r.getLong("request_count") > 0 );
+            // the following are questionable if they belong here
+            Assert.assertEquals("localhost", r.getString("hostname"));
+            Assertions.assertThat(r.getMap("client_options", String.class, String.class))
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> {
+                        assertThat(value.contains(r.getString("driver_name")));
+                    })
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> {
+                        assertThat(value.contains(r.getString("driver_version")));
+                    });
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java]: nit: I will probably write those as follows:
```suggestion
            Assertions.assertThat(r.getMap("client_options", String.class, String.class))
                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_name"))))
                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_version"))));
```

File: test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (bereng) [inline test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java:1]: In `ToolRunner` you have methods to invoke `nodetool` so you could add a test for `nodetool` as well. If that doesn't cut it a dtest could imo.

File: src/java/org/apache/cassandra/transport/ConnectedClient.java [hunk 5]
```diff
@@ -132,6 +140,8 @@ private SslHandler sslHandler()
                            .put(ADDRESS, remoteAddress().toString())
                            .put(USER, username().orElse(UNDEFINED))
                            .put(VERSION, String.valueOf(protocolVersion()))
+                           .put(CLIENT_OPTIONS, Joiner.on(',').withKeyValueSeparator('=').join(
```
Comment (bereng) [inline src/java/org/apache/cassandra/transport/ConnectedClient.java]: I would join on `, `. The extra space makes things easier to read on the eyes.

File: test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java (hunk 4 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java:1]: I wanted to add more info. If you look at `ToolRunner.invokeNodetool()` callers you'll get plenty of examples of such tests. You will see there are methods to check return codes, std output, errors if any are expected, etc. `TableStatsTest` could be used as a guide. But feel free to do your take on it.

The important bit here is that test will check i.e.: The tool doesn't explode when using that option, the new info is indeed returned, return codes are as expected, tests things from the same top-level point of view a user will, etc.

Please let me know if you find any trouble getting it to work and I can happily lend a hand.

File: test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                        "                [--by-protocol] [--clear-history]\n" +
+                        "\n" +
+                        "OPTIONS\n" +
+                        "        --all\n" +
+                        "            Lists all connections\n" +
+                        "\n" +
+                        "        --by-protocol\n" +
+                        "            Lists most recent client connections by protocol version\n" +
+                        "\n" +
+                        "        --clear-history\n" +
```
Comment (bereng) [inline test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java:82]: Nice test!. If you add a test case for `--clear-history`, despite not being directly related to the scope of this ticket, you'd have a complete `clientstats` test class which I think it's a quick win. Wdyt?

**Comment by bereng** [comment]
Hi I was wondering about the revert of the test?

**Comment by rtib** [comment]
Just moved that into another branch. I'd like to fix it, but currently I didn't found how.

File: src/java/org/apache/cassandra/transport/ConnectedClient.java [hunk 7]
```diff
@@ -132,6 +140,8 @@ private SslHandler sslHandler()
                            .put(ADDRESS, remoteAddress().toString())
                            .put(USER, username().orElse(UNDEFINED))
                            .put(VERSION, String.valueOf(protocolVersion()))
+                           .put(CLIENT_OPTIONS, Joiner.on(", ").withKeyValueSeparator('=').join(
+                                   clientOptions().orElse(Collections.<String, String>emptyMap())))
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/transport/ConnectedClient.java]:
```suggestion
                           .put(CLIENT_OPTIONS, Joiner.on(", ")
                                                                             .withKeyValueSeparator("=")
                                                                             .join(clientOptions().orElse(Collections.emptyMap())))
```

File: test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    
+    @Test
+    public void testSelectAll() throws Throwable
+    {
+        ResultSet result = executeNet("SELECT * FROM vts.clients");
+        
+        for (Row r : result)
+        {
+            Assert.assertEquals(InetAddress.getLoopbackAddress(), r.getInet("address"));
+            Assert.assertNotNull(r.getInt("port"));
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java]: Seems like `getInt` already checks for null?

File: test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java [hunk 9]
```diff
@@ -67,12 +67,8 @@ public void testSelectAll() throws Throwable
...
             Assert.assertEquals("localhost", r.getString("hostname"));
             Assertions.assertThat(r.getMap("client_options", String.class, String.class))
-                    .hasEntrySatisfying("DRIVER_VERSION", value -> {
-                        assertThat(value.contains(r.getString("driver_name")));
-                    })
-                    .hasEntrySatisfying("DRIVER_VERSION", value -> {
-                        assertThat(value.contains(r.getString("driver_version")));
-                    });
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_name"))))
+                    .hasEntrySatisfying("DRIVER_VERSION", value -> assertThat(value.contains(r.getString("driver_version"))));
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/db/virtual/ClientsTableTest.java]: nit: formatting is a bit off

File: test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        startJMXServer();
+        daemon.activate();
+        daemon.startNativeTransport();
+        StorageService.instance.registerDaemon(daemon);
+    }
+    
+    @Before
+    public void config() throws Throwable
+    {
+        ResultSet result = executeNet("select release_version from system.local");
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java]: This seems unused?

File: test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        ToolRunner.ToolResult tool = ToolRunner.invokeNodetool("clientstats");
+        tool.assertOnCleanExit();
+        String stdout = tool.getStdout();
+        assertThat(stdout).contains("Total connected clients: 2");
+        assertThat(stdout).contains("User      Connections");
+        assertThat(stdout).contains("anonymous 2");
+    }
+    
+    @Test
+    public void testClientStatsByProtocol() throws Throwable
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/tools/nodetool/ClientStatsTest.java]: These `Throwables` are not needed

**Review by ekaterinadimitrova2 — COMMENTED** [review]
LGTM, I added a few comments that I also suggest on this branch https://github.com/apache/cassandra/compare/trunk...ekaterinadimitrova2:CASSANDRA-16378

**Pull request closed without merging**